pil-analyzer/tests/types.rs

```diff
@@ -614,3 +669,88 @@ fn new_fixed_column_wrong_type() {
     "#;
     type_check(input, &[]);
 }
+
+#[test]
+fn impl_type_resolution() {
+    let input = "
+    namespace std::convert(4);
+        let fe = || fe();
+    namespace F(4);
+    
+        trait Add<T> {
+            add: T, T -> T,
+        }
+
+        impl Add<fe> {
+            add: |a, b| a + b,
+        }
+
+        impl Add<int> {
+            add: |a, b| a + b,
+        }
+
+        let r: int = Add::add(3, 4);
+    ";
+    type_check(input, &[("F::r", "", "int")]);
+}
+
+#[test]
+fn trait_multi_generics() {
+    let input = "
+    trait ToTuple<S, I> {
+        get: S -> (S, I),
+    }
+    impl ToTuple<int, (int, int)> {
+        get: |n| (n, (1, n+2)),
+    }
+    let r: (int, (int, int)) = ToTuple::get(3);
+    ";
+    type_check(input, &[("r", "", "(int, (int, int))")]);
```

Comment:
It is useless to check for the type (i.e. to list `r` in the array you pass to `type_check`), since you already fully specified the type. You can just call the function with an empty array.